### PR TITLE
T5 pipeline parallel

### PR DESCRIPTION
**Pull request by MaximumEntropy**
# What does this PR do ?

Adds pipeline parallel training support to T5.

**Collection**: NLP

# Changelog 
- TBD

# Usage

Set `pipeline_model_parallel_size=2/4` in `megatron_t5_config.yaml`.

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py [hunk 1]
```diff
@@ -113,102 +116,156 @@ def __init__(
         self.post_process = post_process
         self.fp16_cross_entropy = fp16_cross_entropy
         self.precision = precision
+        self.add_encoder = add_encoder
```
Comment (michalivne) [inline nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py:119]: Are we preparing this class to support encoder or decoder only models (so we can disable either)?

File: nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py [hunk 2]
```diff
@@ -226,55 +283,60 @@ def forward(
         """
         Return value is per token / per dimension (i.e., non collapsed loss value)
```
Comment (michalivne) [inline nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py:284]: A general question: does the PP limit returning a dict? Does this means that we need to run method multiple times to get more than a single value?

**Review by michalivne — COMMENTED** [review]
Fantastic PR! Great job with minimizing code duplication to the bare minimum. See comments.

File: Jenkinsfile [hunk 3]
```diff
@@ -2286,7 +2286,7 @@ pipeline {
         trainer.log_every_n_steps=1 \
         trainer.val_check_interval=10 \
         trainer.limit_val_batches=2 \
-        trainer.accumulate_grad_batches=2 \
+        trainer.accumulate_grad_batches=1 \
```
[the remark this answers is not in the document]
Comment (MaximumEntropy) [reply Jenkinsfile:2289]: PTL-based grad accumulation will not work when using apex fwd/bwd, even when using TP=1 and PP=1

**Review by michalivne — APPROVED** [review]
Nice work with this PR! Very exciting.

File: examples/nlp/language_modeling/conf/megatron_t5_config_finetune_glue_mnli.yaml [hunk 4]
```diff
@@ -36,29 +36,36 @@ exp_manager:
     save_best_model: True
 
 model:
-  restore_from_path: ??? # Path to a trained T5 .nemo file
+  restore_from_path: /raid/Data/NMT/Models/mt5_220m/refactored/seq_1024_drop0/megatron_t5--val_loss=1.45-step=610999-consumed_samples=625643520.0.nemo # Path to a trained T5 .nemo file
```
[the remark this answers is not in the document]
Comment (MaximumEntropy) [reply examples/nlp/language_modeling/conf/megatron_t5_config_finetune_glue_mnli.yaml]: I will revert this. This is just for debugging.

File: nemo/collections/nlp/models/language_modeling/megatron_glue_model.py [hunk 5]
```diff
@@ -158,37 +186,128 @@ def __init__(self, cfg: DictConfig, trainer: Trainer):
...
+        _reconfigure_microbatch_calculator(
+            rank=app_state.global_rank,
+            rampup_batch_size=None,
+            global_batch_size=self.cfg.data.train_ds.global_batch_size,
+            micro_batch_size=self.cfg.data.train_ds.micro_batch_size,
+            data_parallel_size=parallel_state.get_data_parallel_world_size(),
+        )
+        return super().on_validation_model_train()
+
     def training_step(self, batch, batch_idx):
```
[the remark this answers is not in the document]
Comment (MaximumEntropy) [reply nemo/collections/nlp/models/language_modeling/megatron_glue_model.py]: This is now done and I've refactored things.

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py [hunk 6]
```diff
@@ -55,46 +74,29 @@ class MegatronLMEncoderDecoderModel(MegatronBaseModel):
...
-            layernorm_epsilon=cfg.get('layernorm_epsilon', 1e-5),
-            persist_layer_norm=cfg.get('persist_layer_norm', False),
-            bias_gelu_fusion=cfg.get('bias_gelu_fusion', True),
-            masked_softmax_fusion=cfg.get('masked_softmax_fusion', True),
-            onnx_safe=cfg.get('onnx_safe', False),
-            activation=cfg.get('activation', 'gelu'),
-        )
+        # TODO: Not sure how to use lists of modules with PTL.
+        # This means we can only use pipeline parallelism without the interleaved schedule.
+        self.enc_dec_model = build_model(
```
[the remark this answers is not in the document]
Comment (MaximumEntropy) [reply nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:95]: Yup!

File: reinstall.sh [hunk 7]
```diff
@@ -28,7 +28,7 @@ else
   then
     NUMBA_VERSION=0.55
     echo 'Installing numba=='${NUMBA_VERSION}
-    conda install -y -c conda-forge numba==${NUMBA_VERSION}
+    # conda install -y -c conda-forge numba==${NUMBA_VERSION}
```
[the remark this answers is not in the document]
Comment (MaximumEntropy) [reply reinstall.sh:31]: I will revert this. This is temporary.

File: reinstall.sh [hunk 8]
```diff
@@ -44,6 +44,6 @@ else
 fi
 
 echo 'Installing additional nemo_text_processing dependency'
-bash nemo_text_processing/setup.sh > /dev/null 2>&1 && echo "nemo_text_processing installed!" || echo "nemo_text_processing could not be installed!"
+# bash nemo_text_processing/setup.sh > /dev/null 2>&1 && echo "nemo_text_processing installed!" || echo "nemo_text_processing could not be installed!"
```
[the remark this answers is not in the document]
Comment (MaximumEntropy) [reply reinstall.sh:47]: Same as Numba. Will revert this for faster debugging.

File: nemo/collections/nlp/parts/nlp_overrides.py [hunk 9]
```diff
@@ -563,3 +574,69 @@ def forward_context(self) -> Generator[None, None, None]:
             yield
         finally:
             pass
+
+
+class NLPDataConnector(DataConnector):
```
[the remark this answers is not in the document]
Comment (MaximumEntropy) [reply nemo/collections/nlp/parts/nlp_overrides.py:579]: We have a fix for GPT that does not need this. I will first merge stuff with this first and then send a follow-up PR with the GPT fix.

File: nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py (hunk 1 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py:119]: No, this is PP specific since certain GPUs may have encoder only or decoder only.

File: nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py (hunk 2 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py:284]: Only torch tensors can be communicated across the pipeline. Only the last pipeline stage can return a dict (since it does not need to be communicated forward).

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 10]
```diff
@@ -34,68 +35,35 @@
...
-                )
-            else:
-                weight_decay_params['params'].extend(
-                    [p for n, p in list(module_._parameters.items()) if p is not None and n != 'bias']
-                )
-                no_weight_decay_params['params'].extend(
-                    [p for n, p in list(module_._parameters.items()) if p is not None and n == 'bias']
-                )
-
-    return weight_decay_params, no_weight_decay_params
```
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:95]: Why is this deleted?

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 11]
```diff
@@ -34,68 +35,35 @@
...
-
-
-class MegatronGPTModel(NLPModel, TextGeneration):
```
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:98]: Why is this deleted?

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 12]
```diff
@@ -249,7 +217,7 @@ def forward(self, tokens, text_position_ids, attention_mask, labels, prompt_ids=
 
     def setup_optimizer_param_groups(self):
         """ModelPT override. Optimizer will get self._optimizer_param_groups"""
-        self._optimizer_param_groups = _get_params_for_weight_decay_optimization([self.model])
```
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:252]: Is this right?

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py (hunk 10 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:95]: It moved to a different file and is imported from there.

Comment (MaximumEntropy) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:95]: This is because I needed it for T5 as well.

Comment (ericharper) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:95]: Thanks.

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py (hunk 12 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:252]: Yeah I just removed the _ when renaming the function.

**Review by ericharper — DISMISSED** [review]
LGTM. Thanks!

**Review by ericharper — APPROVED** [review]
LGTM.
